2-Draft-Accepted/RFCxxxx-PowerShellGet-3.0.md

```diff
@@ -0,0 +1,242 @@
+
+With support of the generic `PSResource`, this means we can also find and
+install arbitrary nupkgs that may only contain an assembly the user wants to
+use in their script.
+
+If the local cache does not exist, the cmdlet will call `Update-PSResourceCache`
+first and then attempt a search.
+However, this cmdlet does not update the cache if one exists.
+
+### Installing resources
+
+`Install-PSResource` will only work for modules.
+A `-AllowPrerelease` switch allows installing prerelease versions.
+Other types will use `Save-PSResource` (see below).
+A `-Repository` parameter accepts a specific repository name.
+If `-Repository` is not specified and the same resource is found in multiple
+repositories, then the resource is automatically installed quietly from the
+trusted repository with the highest version matching the `-Version` parameter
+(if specified, otherwise newest non-prerelease version unless `-AllowPrerelease`
+is used).
+If there are no trusted repositories matching the query, then the newest version
+fulfilling the query will be prompted to be installed.
+If there are multiple repositories with the same trust level containing the same
+version, the first one is used.
+
+`-AllowUntrusted` can be used to suppress being prompted for untrusted sources.
+`-AllowDifferentPublisher` can be used to suppress being prompted if the publisher
+of the module is different from the currently installed version.
+`-AllowReinstall` can be used to re-install a module.
+
+A `-Quiet` switch will suppress progress information.
+
+### Dependencies and version management
+
+`Install-PSResource` can accept a path to a psd1 file (using `-RequiredModulesFile`)
+or a hashtable (using `-RequiredModules`) where the key is the module name and the
+value is either the required version specified using Nuget version range syntax or
+a hash table where `repository` is set to the URL of the repository and
+`version` contains the [Nuget version range syntax](https://docs.microsoft.com/en-us/nuget/reference/package-versioning#version-ranges-and-wildcards).
+
+```powershell
+Install-PSResource -RequiredModules @{
+  "Configuration" = "[1.3.1,2.0)"
+  "Pester"        = @{
+    version = "[4.4.2,4.7.0]"
+    repository = "https://www.powershellgallery.com"
+  }
+}
+```
+
+### Saving resources
+
+With the removal of PackageManagement, there is still a need to support saving
+arbitrary nupkgs (assemblies) used for scripts.
+
+`Save-PSResource -Type Library` will download nupkgs that have a `lib` folder.
+A `-AllowPrerelease` switch allows saving prerelease versions.
+The dependent native library in `runtimes` matching the current system runtime
+will be copied to the same destination.
+
+This cmdlet has the same parameters as `Install-PSResource` with the addition
+of mandatory `-Path` or `-LiteralPath` parameters.
+
+### Updating resources
+
+`Update-PSResource` will update all resources to most recent minor version by
```

Comment:
Will `Update-PSResource` address [PowerShell/PowerShellGet#130](https://github.com/PowerShell/PowerShellGet/issues/130) where `Update-Module` is actually doing a side-by-side install and not removing the previous version?

Reply:
The problem there is that you cannot tell whether it is safe to remove the previous version.

Reply:
Why not at least let the user decide with a new switch, like either just `-Force` or `-ForceUpdate` if you don't want to convolute "force"?

Reply:
Yes, you can by checking the all installed modules `RequiredModules` and leaving the previous version installed if the requirement is not met by the newly installed version. It's the same logic that would need to be implemented for [PowerShell/PowerShellGet#114](https://github.com/PowerShell/PowerShellGet/issues/114).

Reply:
Agree with the concept, but will probably defer to later to implement.  Will need to update this RFC indicating how the changes will be staged as I expect we can't get it all in time for PS7.

Reply:
@tnieto88 other modules aren't the only things with dependencies. Any script file, anywhere on the system, can have a hard-coded version dependency.

Reply:
@Jaykul Yeah, you're right. I can't believe I forgot that scenario, must have wrote my reply too late last night.

Reply:
If the `Update-PSResource` were to remove existing versions too (however it's implemented) then it would need to run a scan on any scripts / modules to check for dependencies in them as to warn about their need to update them too as not to break existing functionality due to module version being a hard coded dependency, which should be implemented in all scripts/modules for defensive purposes.

Whilst this functionality would be nice to have in PSGet it would be better used as part of a mature module/script deployment pipeline that tests scripts/modules for any hidden breakages caused by a dependency update and if no breakages are found can then auto push the updates to all applicable nodes. 
The single user/single machine usecase may be realistic to code for if all scripts where installed with `Install-Script` but the multi user/multi machine usecase is a much more difficult use case that needs to be carefully considered here

Reply:
I've been thinking about this. 

I think we **really** need to _take as many hints as possible from the package managers that have come before us_ and in this case, that probably means: don't worry so much about things which might have been written by the user using packages that were installed _incidentally_ as dependencies. If they wrote them, they'll be able to figure it out.

That is (if I understand correctly) if we track which resources were _intentionally_ installed, then if a resource wasn't installed _intentionally_, and we're uninstalling the last resource that depends on it, we can remove it. If users decide to use a resource that got installed as a dependency, all they have to do is explicitly install/upgrade it to update the tracking data so it's tracked as a individually installed package....

Reply:
Out of scope of this RFC, but something we can look into in the future as it's additive.